### PR TITLE
[Small Feature] Classification: Neural Network and parallelized Random Forest

### PR DESCRIPTION
**Pull request by sgiraudot**
## Rationale

The Classification package is continuously evolving: these are the latest features, whose main interests are a new classifier (Neural Network) and the parallelization of the training algorithm of Random Forest.

## Summary of API changes

- Add an experimental classifier `TensorFlow_neural_network_classifier` using the [TensorFlow](https://www.tensorflow.org/) library.
- Parallelize the training algorithm `ETHZ_random_forest_classifier`.
- Add a constructor to copy this random forest classifier
- Add a method to get the feature usage in random forest (number of times a feature is used as a splitter in the trees' nodes)
- Add 3 new geometric features, `Height_above`, `Height_below` and `Vertical_range`.
- Add a (non-document so far) variant of `classify()` to get the detailed output (probabilities for each label)
- Add warnings in the documentations of feature generators to prevent the user from deleting them after generating the features (features _need_ the generator to be still a valid reference afterwards)

In addition, many enhancements and corretions were made, and the
plug-in was improved once again.

## License and copyright ownership

Inria/GeometryFactory (no change)

## CHANGES.md

```markdown
### Classification

-   Added a new experimental classifier
    `TensorFlow::Neural_network_classifier`.

-   For uniformity, `ETHZ_random_forest_classifier` is renamed
    `ETHZ::Random_forest_classifier` and `OpenCV_random_forest_classifier`
    is renamed `OpenCV::Random_forest_classifier`.

-   The training algorithm of `ETHZ::Random_forest_classifier` was
    parallelized.

-   Added a constructor to copy a `ETHZ::Random_forest_classifier` using a
    different data set as input.

-   Added 3 new geometric features, `Height_above`, `Height_below` and
    `Vertical_range`.
```

## Status

-   Developed and locally tested (GNU/Linux)
-   (Empty) small feature page: [Classification: Neural Network and parallelized Random Forest](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Classification:_Neural_Network_and_parallelized_Random_Forest)
- Pre-approved 2019-01-25 by @sloriot 

## Submission

-   [**Updated User Manual**](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/index.html#Chapter_Classification):
-   [**Updated Reference Manual**](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/group__PkgClassificationRef.html):
    -  [`TensorFlow_neural_network_classifier`](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/classCGAL_1_1Classification_1_1TensorFlow__neural__network__classifier.html)
    -  [`ETHZ_random_forest_classifier` (parallelized traing and `get_feature_usage()` method](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/classCGAL_1_1Classification_1_1ETHZ__random__forest__classifier.html)
    -  [`Height_above`](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/classCGAL_1_1Classification_1_1Feature_1_1Height__above.html)
    -  [`Height_below`](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/classCGAL_1_1Classification_1_1Feature_1_1Height__below.html)
    -  [`Vertical_range`](https://cgal.geometryfactory.com/~sgiraudot/Classification_3.0_v1/Classification/classCGAL_1_1Classification_1_1Feature_1_1Vertical__range.html)

### CODE REVIEW

**Comment by maxGimeno** [comment]
Travis is not happy because there are errors in the doc : 
In file included from /home/travis/build/CGAL/cgal/Installation/../Classification/include/CGAL/Classification/ETHZ_random_forest_classifier.h:44:
/home/travis/build/CGAL/cgal/Classification/include/CGAL/Classification/internal/auxiliary/random-forest/forest.hpp:178:76: error: use of undeclared identifier 'Parallel_tag'
        CGAL_static_assertion_msg (!(boost::is_convertible<ConcurrencyTag, Parallel_tag>::value),
                                                                           ^
/home/travis/build/CGAL/cgal/Classification/include/CGAL/Classification/internal/auxiliary/random-forest/forest.hpp:178:91: error: no type named 'value' in the global namespace
        CGAL_static_assertion_msg (!(boost::is_convertible<ConcurrencyTag, Parallel_tag>::value),
                                                                                        ~~^
/home/travis/build/CGAL/cgal/Installation/../Classification/include/CGAL/Classification/ETHZ_random_forest_classifier.h:136:17: error: no member named 'Sequential_tag' in namespace 'CGAL'
    train<CGAL::Sequential_tag>(ground_truth, reset_trees, num_trees, max_depth);
          ~~~~~~^
/home/travis/build/CGAL/cgal/Installation/../Classification/include/CGAL/Classification/ETHZ_random_forest_classifier.h:201:5: error: use of undeclared identifier 'CGAL_CLASSIFICATION_CERR'
    CGAL_CLASSIFICATION_CERR << "Using " << gt.size() << " inliers" << std::endl;
    ^
4 errors generated.
CMake Error at /home/travis/build/CGAL/cgal/Installation/cmake/modules/run_cmd_redirection_cerr.cmake:32 

/travis/build/CGAL/cgal/Installation/../Classification/include/CGAL/Classification/TensorFlow_neural_network_classifier.h:29:10: fatal error: 'tensorflow/cc/client/client_session.h' file not found
#include <tensorflow/cc/client/client_session.h>
         ^

**Comment by sgiraudot** [comment]
I fixed all the errors except for the last one.

For the last one, I guess Travis should not check `TensorFlow_neural_network_classifier` if TensorFlow is not installed, right?

**Comment by maxGimeno** [comment]
A possibility would be to add the problematic header in the list at 
CGAL/Installation/cmake/modules/list_of_whitelisted_headers.cmake.
Maybe add a cmake condition to only add it if travis is running, or not. I don't know.

**Comment by maxGimeno** [comment]
and https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-72.shtml#Polyhedron_Demo

For the polyhedron demo errors, compile in NOT header only and you should reproduce the missing headers.

**Comment by danston** [comment]
I think, this version of the package is very clean, so I do not have many comments. All the changes are clear. I only found a few typos, which would be good to fix for better readability.

**User manual:**
the point and local lowest point - the point and the local lowest point (I would add "the" article to local lowest point)

**Ref manual:**
_TensorFlow classifier:_ size of the random subset of inliers uses for optimizing at each iteration -> size of the random subset of inliers used for optimizing at each iteration (I think you meant used instead of uses)
_TensorFlow classifier:_ discarding all previous training result -> discarding all previous training results (I think it is better to use plural for results)

_get_feature_usage():_ in the foret that uses it as a split criterion -> in the forest that uses it as a split criterion (s is missing in forest)

_Height_above:_ between the maxium height -> between the maximum height (m is missing in maximum)
_Height_above:_ Its default name is "height_after" -> maybe it is better to change its default name to "height_above"?

_Vertical_range:_ Feature based on local height distribution -> Feature based on local height distribution. (forgot the dot at the end)

This is it from my side.

**Comment by sgiraudot** [comment]
> _Height_above:_ Its default name is "height_after" -> maybe it is better to change its default name to "height_above"?
That was an auto-complete fail in the doc (the name is indeed `height_after`).

Thanks for the review, I corrected all these mistakes!

**Comment by sloriot** [comment]
Breaks the demo of CGAL-4.14-Ic-76

**Comment by sgiraudot** [comment]
Stupid mistake I did, forgetting to add a file to the branch… It should be fixed now.

**Comment by sloriot** [comment]
Successfully tested in CGAL-4.14-Ic-87

**Comment by maxGimeno** [comment]
Error in travis : /home/travis/build/CGAL/cgal/Installation/../Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h:29:10: fatal error: 'tensorflow/cc/client/client_session.h' file not found
#include <tensorflow/cc/client/client_session.h>

**Comment by sgiraudot** [comment]
I forgot to update the filename that I changed in the whitelisted headers… it should work now.

**Comment by lrineau** [comment]
This PR is waiting for a pre-approval. @sgiraudot, you should try and find a reviewer.

**Comment by lrineau** [comment]
[**@lrineau**](https://github.com/lrineau) commented on [Jan 18, 2019, 4:45 PM GMT+1](https://github.com/CGAL/cgal/pull/3535#issuecomment-455590394 "2019-01-18T15:45:27Z - Replied by Github Reply Comments"):
> This PR is waiting for a pre-approval. @sgiraudot, you should try and find a reviewer.

@sgiraudot ping!

**Comment by sgiraudot** [comment]
@sloriot and @afabri were both reviewers for the original package. Could either one of you do a quick review for this one? Dmitry already gave its input but we need an EB member to validate it (or not).

File: Classification/doc/Classification/Classification.txt [hunk 1]
```diff
@@ -244,6 +249,31 @@ use this classifier. For more details about the algorithm, please refer
...
 
+\subsection Classification_TensorFlow_neural_network TensorFlow Neural Network
+
+\cgal provides [TensorFlow::Neural_network_classifier](@ref CGAL::Classification::TensorFlow::Neural_network_classifier).
+
+It uses the C++ API of the \ref thirdpartyTensorFlow library.
+
+\warning This feature is still experimental: it may not be stable
+and is likely to undergo substantial changes in future releases of
+\cgal. Use at your own risk.
```
Comment (sloriot) [inline Classification/doc/Classification/Classification.txt]: No need for `Use at your own risk`. Instead indicates that API changes will be announced in release notes.

File: Classification/doc/Classification/Classification.txt [hunk 2]
```diff
@@ -478,23 +508,29 @@ The following example:
...
 \subsection Classification_example_opencv_random_forest OpenCV Random Forest
 
-The following example shows how to use the classifier [OpenCV_random_forest_classifier](@ref CGAL::Classification::OpenCV_random_forest_classifier) using an input training set.
+The following example shows how to use the classifier [OpenCV::Random_forest_classifier](@ref CGAL::Classification::OpenCV::Random_forest_classifier) using an input training set.
 
 \cgalExample{Classification/example_opencv_random_forest.cpp}
 
+\subsection Classification_example_tensorflow_neural_network TensorFlow Neural Network
+
+The following example shows how to use the classifier [TensorFlow::Neural_network_classifier](@ref CGAL::Classification::TensorFlow::Neural_network_classifier) using an input training set.
```
Comment (sloriot) [inline Classification/doc/Classification/Classification.txt]: backticks missing

File: Classification/doc/Classification/Classification.txt [hunk 3]
```diff
@@ -478,23 +508,29 @@ The following example:
...
+\cgalExample{Classification/example_tensorflow_neural_network.cpp}
+
 \subsection Classification_example_mesh Mesh Classification
 
 The following example:
 
 - reads a mesh in OFF format;
 - automatically generates features on 5 scales;
-- loads a configuration file for classifier [ETHZ_random_forest_classifier](@ref CGAL::Classification::ETHZ_random_forest_classifier);
+- loads a configuration file for classifier [ETHZ::Random_forest_classifier](@ref CGAL::Classification::ETHZ::Random_forest_classifier);
```
Comment (sloriot) [inline Classification/doc/Classification/Classification.txt]: backticks

File: Classification/doc/Classification/Classification.txt [hunk 4]
```diff
@@ -509,15 +545,15 @@ The following example:
 - detects plane using the algorithm `CGAL::Shape_detection_3::Region_growing`;
 - creates [Cluster](@ref CGAL::Classification::Cluster) objects from these detected planes;
 - computes cluster features from the pointwise features;
-- loads a configuration file for classifier [ETHZ_random_forest_classifier](@ref CGAL::Classification::ETHZ_random_forest_classifier);
+- loads a configuration file for classifier [ETHZ::Random_forest_classifier](@ref CGAL::Classification::ETHZ::Random_forest_classifier);
```
Comment (sloriot) [inline Classification/doc/Classification/Classification.txt]: backticks

File: Classification/include/CGAL/Classification/ETHZ/Random_forest_classifier.h [hunk 5]
```diff
@@ -195,10 +241,46 @@ class ETHZ_random_forest_classifier
...
+    Each tree of the random forest recursively splits the training
+    data set using at each node one of the input features. This method
+    counts, for each feature, how many times it was selected by the
+    training algorithm as a split criterion.
+
+    This method allows to evaluate how useful a feature was with
+    respect to a training set: if a feature is used a lot, that means
+    that it has a strong discriminative power with respect to how the
+    labels are represented by the feature set; on the contrary, if a
+    feature is not used very often, it's discriminative power is
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/ETHZ/Random_forest_classifier.h]: its

File: Classification/include/CGAL/Classification/Feature/Height_above.h [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+namespace CGAL {
+
+namespace Classification {
+
+namespace Feature {
+
+  /*!
+    \ingroup PkgClassificationFeatures
+
+    %Feature based on local height distribution This feature computes
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/Feature/Height_above.h]: missing dot?

File: Classification/include/CGAL/Classification/Feature/Height_below.h [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+namespace CGAL {
+
+namespace Classification {
+
+namespace Feature {
+
+  /*!
+    \ingroup PkgClassificationFeatures
+
+    %Feature based on local height distribution This feature computes
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/Feature/Height_below.h]: missing .?

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h [hunk 8]
```diff
@@ -0,0 +1,1122 @@
...
+
+  This class provides an interface to a feature-based neural network:
+  a set of features is used as an input layer followed by a
+  user-specified number of hidden layers with a user-specified
+  activation function. The output layer is a softmax layer providing,
+  for each label, the probability that an input item belongs to it.
+
+  \warning This feature is still experimental: it may not be stable
+  and is likely to undergo substantial changes in future releases of
+  \cgal. Use at your own risk.
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: same remark as above.

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h [hunk 9]
```diff
@@ -0,0 +1,1122 @@
...
+
+  \note This class requires the \ref thirdpartyTensorFlow library.
+
+  \tparam ActivationFunction Chosen activation function for the hidden
+  layers. Relu is used as default. The following functions can be used
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: backtick Relu?

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h [hunk 10]
```diff
@@ -0,0 +1,1122 @@
...
+  (please refer to the documentation of TensorFlow for more
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: \thirdparty blabla?

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h [hunk 11]
```diff
@@ -0,0 +1,1122 @@
...
+
+  /// \cond SKIP_IN_MANUAL
+  const Feature_set& features() const { return m_features; }
+  /// \endcond
+  
+  /// \name Constructor
+  /// @{
+  
+  /*!
+    \brief Instantiate the classifier using the sets of `labels` and `features`.
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: instantiates

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h [hunk 12]
```diff
@@ -0,0 +1,1122 @@
...
+    label.
+
+    \param ground_truth vector of label indices. It should contain for
+    each input item, in the same order as the input set, the index of
+    the corresponding label in the `Label_set` provided in the
+    constructor. Input items that do not have a ground truth
+    information should be given the value `-1`.
+
+    \param restart_from_scratch should be set to `false` if the user
+    wants to continue adjusting weights and bias based, and kept to
```
Comment (sloriot) [inline Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: set vs kept, set in both cases?

File: Classification/doc/Classification/Classification.txt (hunk 2 above)
Comment (sgiraudot) [reply Classification/doc/Classification/Classification.txt]: What do you mean? Backticks are used to create autolinks, here I'm creating an explicit link (because I don't want the full name with namespaces to show), so backticks are not needed unless I'm mistaken.

Comment (sloriot) [reply Classification/doc/Classification/Classification.txt]: backtick = typewriting [`toto`](mylink) AFAIK

**Comment by afabri** [comment]
I would like to suggest to make the recommendation which classifier to use before the subsections for each classifier.  Also don't use bold but a \note as bold should be reserved for headings.

File: Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h (hunk 12 above)
Comment (sgiraudot) [reply Classification/include/CGAL/Classification/TensorFlow/Neural_network_classifier.h]: "Kept" was to indicate that it was the default value, but okay.

File: Classification/doc/Classification/Classification.txt (hunk 2 above)
Comment (sloriot) [reply Classification/doc/Classification/Classification.txt]: (you need to click on edit, I can verbatim the toto link :) )

Comment (lrineau) [reply Classification/doc/Classification/Classification.txt]: Yes you can:
```
[`toto`](mylink)
```

Comment (sgiraudot) [reply Classification/doc/Classification/Classification.txt]: It doesn't work, if I do
```
[`toto`](mylink)
```
then it actually shows the backticks around the link in the produced doc. If I do
```
`[toto](mylink)`
```
then it shows the unaltered markdown code…

**Comment by maxGimeno** [comment]
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-4.14-Ic-116.shtml
